app/src/main/java/de/westnordost/streetcomplete/util/CheckIfNight.kt

```diff
@@ -0,0 +1,36 @@
+package de.westnordost.streetcomplete.util
+
+import com.luckycatlabs.sunrisesunset.Zenith
+import com.luckycatlabs.sunrisesunset.calculator.SolarEventCalculator
+import com.luckycatlabs.sunrisesunset.dto.Location
+import de.westnordost.streetcomplete.data.osm.mapdata.LatLon
+import java.time.LocalDate
+import java.time.LocalTime
+import java.time.ZoneId
+import java.time.ZonedDateTime
+import java.util.*
+
+fun localDateToCalendar(localDate: LocalDate): Calendar {
+    val calendar = Calendar.getInstance()
+    calendar.set(localDate.year, localDate.month.value, localDate.dayOfMonth)
+    return calendar
+}
+
+fun isNight(pos: LatLon): Boolean {
```

Comment:
Ok, I now recognize that here is a potential for a performance issue. Let's say you have 1000 quests that shall not be visible during day and they shall now be displayed on the map.

Then, the calculation in `CheckIfNight` is executed 1000 times. I am not sure how expensive the calculation is, but to circumvent that, I suggest the following:

1. `isNight` calculates the sunset/dawn with the given latitude and longitude truncated to two decimals ([precision of ~1km](https://wiki.openstreetmap.org/wiki/Precision_of_coordinates)). That should be enough precision
2. the calculation is cached in a dayTime: Map<LatLon, Pair<LocalTime,LocalTime>>. Only if there is no entry in the map, the calculation is done. LatLon is always the truncated latlon.
3. If `isNight` is going to have a state, it would make sense to make this into a class, or maybe just put the code here in `DayNightQuestFilter`

Reply:
Whilst I fully agree with you, I'm not sure how to go about implementing this

Reply:
Hmm… … how should this feature actually work?

1. When it is detected to be after sunset at the user's current position, all lit quests appear.
    - How I imagined it would be when I read the title.
    - Should be synchronized with using the dark theme, if the "Auto" theme is selected.
    - Simple to implement, cheap to compute.
    - What if we don't have the user's current location?
    - What if the user is looking (say, to see what quests are available) at a position different than their current location? Use gps or screen/scroll position?
2. Lit quests appear if it is past sunset at their current location.
    - How it's currently implemented, I think.
    - Potential performance issues, or more complicated implemented to mitigate them.
    - Might be confusing (hard to figure out when a given lit quest will appear), particularly along with the "available in 3 hours" feature suggested below.
        - Might also be *cool* — seeing lit quests gradually appear as the sun crawls across the map. But would definitely work better if the map visualized the current sun line instead of switching all at once.

Personally I'm more inclined toward solution 1 and simply not worrying about making the app work in those edge cases, since they are not the primary use case of StreetComplete (in-person surveying).

Reply:
@smichel17, if you are going to comment on the code, you should read it first before making assumptions.

Reply:
In the diff, I came across this code:

https://github.com/streetcomplete/StreetComplete/blob/e5c6135d0ad1e30e11f6ad25e3ee5474228eeffd/app/src/main/java/de/westnordost/streetcomplete/data/visiblequests/DayNightQuestFilter.kt#L13-L19

Which seems to be applied per-quest. I admit I have not traced the call stack more than one step, to here:

https://github.com/streetcomplete/StreetComplete/blob/e5c6135d0ad1e30e11f6ad25e3ee5474228eeffd/app/src/main/java/de/westnordost/streetcomplete/data/quest/VisibleQuestsSource.kt#L86-L87

to see if it is called differently, but I don't think my assumption — made *after* reading the code — was unreasonable.

---

It seems like I have missed something that you find obvious, but I'm honestly not sure what it is. Also happy to figure it out myself if you want to point me in the right direction.

Reply:
Alright, maybe a misunderstanding. The way you wrote the other comment sounded like you would comment on the functionality in general as if no code had been written yet.

Reply:
In any case, I'd also prefer 1, but I am not going to reject this PR just because 2 and not 1 has been implemented.

If @TurnrDev or you would like to move to behavior 1 instead, I can give advice how to do that on the technical side.